### PR TITLE
build: enable OSX builds

### LINKED ISSUE #128: OSX support
Get Envoy compiling on OSX. Shim out all linux specific features and either turn them off in OSX build or provide OS shim layer to make them work on BSD.

### PR DESCRIPTION
**Pull request by zuercher**
Copies in bazel/tools/osx for xcode support. Fixes bugs related to ar and strip flags.

Modifies build-related shell scripts to account for flag differences between GNU and
OSX command line tools.

Disables tcmalloc on OSX and removes various linker flags incompatible linker flags.

Partially fixes #128.

### CODE REVIEW

File: bazel/envoy_build_system.bzl [hunk 1]
```diff
@@ -20,12 +20,45 @@ def envoy_copts(repository, test = False):
...
         "//conditions:default": ["-DTCMALLOC"],
     }) + select({
         repository + "//bazel:disable_signal_trace": [],
         "//conditions:default": ["-DENVOY_HANDLE_SIGNALS"],
     })
 
+# Compute the final linkopts based on various options.
+def envoy_linkopts():
+    return select({
+        "//bazel/osx:darwin": [],
```
Comment (htuch) [inline bazel/envoy_build_system.bzl]: How come we aren't providing pertinent link options here on OS X? Can we still get the appropriate build hash and libraries linked statically? I assume so as tests pass, but I'm confused why we don't need to provide explicit link options.

File: bazel/envoy_build_system.bzl [hunk 2]
```diff
@@ -20,12 +20,45 @@ def envoy_copts(repository, test = False):
         repository + "//bazel:dbg_build": ["-ggdb3"],
     }) + select({
         repository + "//bazel:disable_tcmalloc": [],
+        "//bazel/osx:darwin": [],
```
Comment (htuch) [inline bazel/envoy_build_system.bzl]: Is this the permanent plan? Does gperftools support OS X? tcmalloc has some performance advantages in addition to performance insight via the tools.

File: bazel/osx/BUILD [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+exports_files([
+    "xcode_locator.m",
+])
+
+DARWIN_XCODE_LOCATOR_COMPILE_COMMAND = """
+  /usr/bin/xcrun clang -fobjc-arc -framework CoreServices \
+      -framework Foundation -o $@ $<
+"""
+
+genrule(
```
Comment (zuercher) [inline bazel/osx/BUILD]: This combined with the code in #1348 should let you build and test on OSX. By itself it just shouldn't break the linux build, I guess.

I'll try to remove some stuff from bazel/osx again -- I had trouble before that I though was related to sandboxing, but I may have been mistaken.

Will add a note about installing realpath.

Comment (htuch) [inline bazel/osx/BUILD]: I don't think @bazel_tools is defined by Bazel. The only way to do this would be to include the Bazel git tree as a `git_repository` and then reference these files. I think that would be reasonable to do for unchanged subtrees. Looking below, I think it might make sense to do this.

File: tools/git_sha_rewriter.py [hunk 4]
```diff
@@ -107,6 +108,9 @@ def RewriteBinary(path, offset, git5_sha1):
   if len(sys.argv) != 2:
     print 'Usage: %s <Envoy binary path> ' % sys.argv[0]
     sys.exit(1)
+  if platform.system() == 'Darwin':
+    print 'Stamping not supported on OSX'
```
Comment (htuch) [inline tools/git_sha_rewriter.py]: What's the limitation here?

File: bazel/osx/BUILD (hunk 3 above)
Comment (lizan) [inline bazel/osx/BUILD]: It is fine if it doesn't work, assuming it is just a temporary workaround until upstream solves it.

@htuch `@bazel_tools` is definitely defined by bazel itself, try run `bazel build @bazel_tools//tools/cpp:stl` ;)

Comment (htuch) [inline bazel/osx/BUILD]: Oh, neat. I could have sworn that in the past that didn't work. In that case, I agree with what you say, we should use references to @bazel_tools for the unmodified files (where they don't have modified dependencies).

File: bazel/envoy_build_system.bzl (hunk 2 above)
Comment (zuercher) [inline bazel/envoy_build_system.bzl]: I cribbed this from @Reflejo's old branch. I believe he encountered problems with tcmalloc deadlocking (c.f. https://github.com/lyft/envoy/issues/128#issuecomment-281342790).

I'm game to try turning it on.

File: bazel/envoy_build_system.bzl (hunk 1 above)
Comment (zuercher) [inline bazel/envoy_build_system.bzl]: OS X system libraries are distributed exclusively as dynamically loaded libraries. So, all the thirdparty dependencies (e.g. ci/build_container/build_recipes) are linked statically, but the binary links libc++ and libSystem dynamically. That second library dynamically links libraries for libpthread and libuuid.

File: tools/git_sha_rewriter.py (hunk 4 above)
Comment (zuercher) [inline tools/git_sha_rewriter.py]: At a minimum, OS X binaries are not ELF format, so this particular tool will not work. I don't know enough about Mach-O to say whether it's possible to do something similar.

File: bazel/osx/BUILD (hunk 3 above)
Comment (zuercher) [inline bazel/osx/BUILD]: FWIW, I could revert the change that enables the darwin compiler toolchain and turn that on in the PR with all the other necessary changes to get it compiling under OSX.

File: bazel/envoy_build_system.bzl (hunk 1 above)
Comment (zuercher) [inline bazel/envoy_build_system.bzl]: As far as build hash goes, the --build-id flag to `ld` requests create of a ".note.gnu.build-id" ELF note section. Not sure if OS X's Mach-O format supports something similar or not.

**Comment by htuch** [comment]
Right, this is Mach-O. I think we don't need this, it's only something we
added for Lyft. I'd recommend just changing the message to "not supported
with Mach-O" to make clear this is a file format issue, not an OS one.

File: bazel/envoy_build_system.bzl (hunk 1 above)
Comment (zuercher) [inline bazel/envoy_build_system.bzl]: I think one could inject a custom section into the mach-o executable by adding a flag to ld. Something like `-sectcreate __TEXT __build_id file`. And then git_rewrite_sha.py could be modified to update it with the git SHA.

Comment (mattklein123) [inline bazel/envoy_build_system.bzl]: Personally I would just put in a TODO. This is really only useful in a production setting when dealing with core dumps.

File: bazel/envoy_build_system.bzl (hunk 2 above)
Comment (zuercher) [inline bazel/envoy_build_system.bzl]: Seems to work, so I put it back in.

File: bazel/envoy_build_system.bzl (hunk 1 above)
Comment (zuercher) [inline bazel/envoy_build_system.bzl]: Added a TODO.

File: bazel/osx/BUILD (hunk 3 above)
Comment (zuercher) [inline bazel/osx/BUILD]: I was able to remove most of the copied bazel files. I've left the `configure_osx_toolchain` invocation in `cc_configure.bzl`.

File: bazel/envoy_build_system.bzl [hunk 5]
```diff
@@ -26,6 +26,41 @@ def envoy_copts(repository, test = False):
         "//conditions:default": ["-DENVOY_HANDLE_SIGNALS"],
     })
 
+# Compute the final linkopts based on various options.
+def envoy_linkopts():
+    return select({
+        # TODO(zuercher): build id could be supported via "-sectcreate __TEXT __build_id <file>"
+        # The file could should contain the current git SHA (or enough placeholder data to allow
+        # it to be rewritten by tools/git_sha_rewriter.py).
```
Comment (htuch) [inline bazel/envoy_build_system.bzl]: Can you also include a comment on how static/system libs work on OS X?

**Comment by htuch** [comment]
@zuercher Can you file issues with https://github.com/bazelbuild/bazel/issues for the static link issue if you haven't already and add references to that in this commit? Ideally these get fixes one day upstream and we can unwind our hacks. Here's the issue I filed to track the Linux stuff https://github.com/bazelbuild/bazel/issues/2840.

**Comment by zuercher** [comment]
@htuch I apologize. The `ar` bug is fixed in upstream bazel. I thought it hadn't because the incorrect flags I was modifing are still there, they just don't get used for some reason. 

The `strip` bug has already been reported by someone else as https://github.com/bazelbuild/bazel/issues/209. If not for that bug, we could use bazel 0.5.2's configure_osx_toolchain without modification.

I modified the bazel/osx/crosstool bits to use the 0.5.2 CROSSTOOL.tpl with just my fix for stripping and added diff headers referring to the strip bug and the git tag the files came from.

File: bazel/osx/crosstool/CROSSTOOL.tpl [hunk 6]
```diff
@@ -1270,8 +1275,7 @@ toolchain {
     config_name: "c++-link-interface-dynamic-library"
     action_name: "c++-link-interface-dynamic-library"
     tool {
-      tool_path: "wrapped_clang"
-      execution_requirement: "requires-darwin"
+      tool_path: "DUMMY_TOOL"
```
Comment (htuch) [inline bazel/osx/crosstool/CROSSTOOL.tpl]: I'm a bit confused by the introduction of `DUMMY_TOOL` here - it doesn't appear to be in `crosstool.tpl.diff` or in the upstream https://github.com/bazelbuild/bazel/blob/master/tools/osx/crosstool/CROSSTOOL.tpl.

**Comment by htuch** [comment]
@zuercher Sounds good on the upstream story. Could you also leave a comment on https://github.com/bazelbuild/bazel/issues/209 to remind folks this is also affecting Envoy? Thanks.

File: bazel/osx/crosstool/CROSSTOOL.tpl (hunk 6 above)
Comment (zuercher) [inline bazel/osx/crosstool/CROSSTOOL.tpl]: I wanted to submit a simple project as a test case for the ar bug, but after I was unable to reproduce it, I realized that the [0.5.2](https://github.com/bazelbuild/bazel/tree/0.5.2) tag is not bazelbuild/bazel@7b85122c9ed1c5bc32ef9c6ac213910adee735e2, the latter being the SHA from which other stuff was copied.

I think what's happened is that a 0.5.2 release branch was started, and the ar bug was fixed there. Meanwhile in master the DUMMY_TOOL entries were removed (presumably because they weren't being used) and the ar bug was also fixed there, but after 7b85122c. I'm at a loss to explain what actually fixes the bug or why the DUMMY_TOOL entries don't seem to bother anything.

Comment (zuercher) [inline bazel/osx/crosstool/CROSSTOOL.tpl]: And I should add, I copied this CROSSTOOL.tpl from the 0.5.2 tag, hence the older DUMMY_TOOL entries.

Comment (mattklein123) [inline bazel/osx/crosstool/CROSSTOOL.tpl]: @zuercher I don't know any of the details here, but FYI 0.5.3 is out of that changes anything.

**Comment by zuercher** [comment]
So, the CROSSTOOL.tpl from 0.5.2 had didn't work with the just-released 0.5.3. However, the 0.5.3 version (plus the strip fix) works in 0.5.2.

I think all the comments are addressed at this point.
